### DISCUSSION TITLE
Sample audio directly on the onboard ADC

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#11504] [ESP32] [answered]
Hi,
does anybody know if it is possible in Micropython to sample audio directly on an ADC pin (12 bit of an ESP32) at a fixed sample rate and how to to this? I'm not looking for hifi quality, I just want to analyze a raw audio signal without using an external audio ADC. 12 bit is enough but I want to have a fixed sample rate.
Using a timer with interrupts or asyncio, not sure what the best approach would be.
Thnx!

**Comment by jimmo** [comment, marked as answer]
On other ports (e.g. STM32) we support `adc.read_timed` which allows the ADC to be sampled at a fixed rate.

Unfortunately this is not (yet) supported on ESP32.

If you want this to run in the background while the code does other things, your best bet is to use a timer and read from the ADC in the interrupt handler. I recommend pre-allocating the output buffer so that the handler doesn't have to do any allocations. Unfortunately the sample rate will not be terrible high.

If you just want to sample as fast as possible, then calling ADC.read in a loop (optionally with time.sleep_us), but there will be quite a lot of jitter in the sampling interval.

**Reply by saraverbeecke** [reply]
Thanks @jimmo

**Comment by bixb922** [comment]
See also #10561
